python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -14,130 +16,374 @@
 logger = logging.getLogger(__name__)
 
 
-class _DistributeResources:
-    """Generic functionality for resource allocation functions"""
+@PublicAPI(stability="beta")
+class DistributeResources:
```

Comment:
Sorry if I am asking some very basic questions :) 
What is the life cycle of this class? It seems to me this class is re-instantiated and called again and again in `evenly_distribute_cpus_gpus`. So then, what's the point of it being a class? And why the distinction of some logic going to `__init__` while other logic going to `__call__`?

Reply:
This class is a callable that's initialized once and called multiple times. It's a class in order to allow for subclassing and modualrity. If it was a simple function, it would be necessary to duplicate a lot of code for other allocation strategies. Furthermore, making it a class allows for some options to be specified during initialization.

Reply:
Maybe an example can help me understand here. By reading the code that is presented here, this class is instantiated multiple times and invoked through __call__ method exactly the same number of times in `evenly_distribute_cpus_gpus`.

Reply:
See the conversation above.